src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,309 @@
+
+        min J(mu), mu in C
+
+    for an output functional depending on a box-constrained `mu` using the BFGS method.
+
+    Parameters
+    ----------
+    model
+        The |Model| used for the optimization.
+    parameter_space
+        The |ParameterSpace| for enforcing the box constraints on the parameter `mu`.
+    initial_guess
+        If not `None`, a |Mu| instance of length 1 containing an initial guess for the
+        solution `mu`.
+    miniter
+        Minimum amount of iterations to perform.
+    maxiter
+        Fail if the iteration count reaches this value without converging.
+    atol
+        Finish when the absolute error measure is below this threshold.
+    tol_sub
+        Finish when the clipped parameter error measure is below this threshold.
+    line_search_params
+        Dictionary of additional parameters passed to the line search method.
+    stagnation_window
+        Finish when the parameter update has been stagnating within a tolerance of
+        `stagnation_threshold` during the last `stagnation_window` iterations.
+    stagnation_threshold
+        See `stagnation_window`.
+    error_aware
+        If `True`, perform an additional error aware check during the line search phase.
+        Intended for use with the trust region algorithm. Requires the parameters `beta`
+        and `radius` to be set.
+    beta
+        Intended for use with the trust region algorithm. Indicates the factor for checking
+        if the current value is close to the trust region boundary. See `error_aware`.
+    radius
+        Intended for use with the trust region algorithm. Indicates the radius of the
+        current trust region. See `error_aware`.
+    return_stages
+        If `True`, return a `list` of the intermediate parameter values of `mu` after
+        each iteration.
+
+    Returns
+    -------
+    mu
+        |Numpy array| containing the computed parameter.
+    data
+        Dict containing the following fields:
+
+            :mus:             `list` of parameters after each iteration.
+            :mu_norms:        |NumPy array| of the parameter norms after each iteration.
+            :update_norms:    |NumPy array| of the norms of the update vectors for each iteration.
+            :stages:          See `return_stages`.
+
+    Raises
+    ------
+    BFGSError
+        Raised if the BFGS algorithm failed to converge.
+    """
+    logger = getLogger('pymor.algorithms.bfgs')
+
+    data = {}
+
+    assert model.output_functional is not None
+    output = lambda m: model.output(m)[0, 0]
+
+    if initial_guess is None:
+        initial_guess = parameter_space.sample_randomly(1)[0]
+        mu = initial_guess.to_numpy()
+    else:
+        mu = initial_guess.to_numpy() if isinstance(initial_guess, Mu) else initial_guess
+
+    if error_aware:
+        assert beta is not None and radius is not None
+
+        def error_aware_line_search_criterion(f, starting_point, alpha, direction, grad, slope):
```

Comment:
I understand that this is the most flexible way of allowing for an additional line search criterion. In my opinion, however, it should be enough to only have `new_mu` and `current_value`, (which is available in the line search) as an argument.